File: src/main/java/com/freya02/docs/data/BaseDoc.java
package com.freya02.docs.data;

import com.freya02.bot.utils.HTMLElement;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.EnumSet;
import java.util.List;

public abstract class BaseDoc {
	@NotNull
	public abstract String getEffectiveURL();

	@Nullable
	public abstract HTMLElement getDescriptionElement();

	@Nullable
	public abstract HTMLElement getDeprecationElement();

	@NotNull
	protected abstract DetailToElementsMap getDetailToElementsMap();

	@NotNull
	public List<DocDetail> getDetails(EnumSet<DocDetailType> includedTypes) {
		final List<DocDetail> details = new ArrayList<>();

		for (DocDetailType detailType : DocDetailType.values()) {
			if (includedTypes.contains(detailType)) {
				final DocDetail detail = getDetailToElementsMap().getDetail(detailType);
				if (detail == null) continue;

				details.add(detail);
			}
		}

		return details;
	}
}

File: src/main/kotlin/com/freya02/docs/data/BaseDoc.kt
package com.freya02.docs.data

import com.freya02.docs.HTMLElement
import com.freya02.docs.HTMLElementList
import java.util.*

abstract class BaseDoc {
    abstract val effectiveURL: String
    abstract val descriptionElements: HTMLElementList
    abstract val deprecationElement: HTMLElement?
    abstract val seeAlso: SeeAlso?

    abstract val className: String
    /** method(Type1, Type2) */
    abstract val identifier: String?
    /** method */
    abstract val identifierNoArgs: String?
    /** method(Type name, name2) */
    abstract val humanIdentifier: String?
    abstract fun toHumanClassIdentifier(className: String): String?

    protected abstract val detailToElementsMap: DetailToElementsMap

    fun getDetails(includedTypes: EnumSet<DocDetailType>): List<DocDetail> =
        DocDetailType.values()
            .filter { it in includedTypes }
            .mapNotNull { detailType -> detailToElementsMap.getDetail(detailType) }
}